### PR TITLE
[WIP] Multi-agent training

### PR DESCRIPTION
**Pull request by lmBored**
Features:
+ This contains the changes from #659 
+ Utilized Pettingzoo + Benchmarl, with concurrent environment workers that collect observations
+ Each game instances is a thread; same game instances are managed by a "collector" process (pipes); all processes are managed by 1 "worker"
+ Architecture is a simple feedforward neural network
+ Only 3 algorithms are supported in the example script: MAPPO, QMIX, MASAC
+ 3 scenarios are supported
+ Default hyperparameters are "tuned"

Roadmap:
1. Add documentations
2. Add actual cooperative scenarios and reward shaping
3. Refactor as Python API
3.1. Generalized support for benchmarl config

P.s. An example run command is at the top of `examples/python/pettingzoo_learning.py`

### CODE REVIEW

File: src/lib_python/CMakeLists.txt [hunk 1]
```diff
@@ -80,6 +80,7 @@ add_custom_target(assemble_package ALL
     COMMAND ${CMAKE_COMMAND} -E copy ${VIZDOOM_SRC_DIR}/freedoom2.wad ${VIZDOOM_PYTHON_PACKAGE_DIR}
     COMMAND ${CMAKE_COMMAND} -E copy_directory ${VIZDOOM_ROOT_DIR}/scenarios ${VIZDOOM_PYTHON_PACKAGE_DIR}/scenarios
     COMMAND ${CMAKE_COMMAND} -E copy_directory ${VIZDOOM_ROOT_DIR}/gymnasium_wrapper ${VIZDOOM_PYTHON_PACKAGE_DIR}/gymnasium_wrapper
+    COMMAND ${CMAKE_COMMAND} -E copy_directory ${VIZDOOM_ROOT_DIR}/gymnasium_wrapper ${VIZDOOM_PYTHON_PACKAGE_DIR}/pettingzoo_wrapper
```
Comment (mwydmuch) [inline src/lib_python/CMakeLists.txt]: Should be
```
    COMMAND ${CMAKE_COMMAND} -E copy_directory ${VIZDOOM_ROOT_DIR}/pettingzoo_wrapper ${VIZDOOM_PYTHON_PACKAGE_DIR}/pettingzoo_wrapper
```

File: pettingzoo_wrapper/base_env_common.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+from typing import Any
+
+import cv2
+import numpy as np
+import pygame
+from gymnasium import spaces
+from pettingzoo import ParallelEnv
+
+import vizdoom as vzd
+from pettingzoo_wrapper.utils import discover_buttons, get_screen_resolution, parse_hw
```
Comment (mwydmuch) [inline pettingzoo_wrapper/base_env_common.py:16]: All these imports should start with `.`: `pettingzoo_wrapper.utils` -> `.utils`, otherwise it won't work after the ViZDoom package is installed.

File: pettingzoo_wrapper/observation_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        channels = 1 if len(sample_shape) < 3 else sample_shape[-1]
+        self._observation_space = spaces.Box(
+            low=sample_space.low.flat[0],
+            high=sample_space.high.flat[0],
+            shape=(self.height, self.width, channels),
+            dtype=sample_space.dtype,
+        )
+
+    def _resize_obs(self, obs: np.ndarray) -> np.ndarray:
+        resized = cv2.resize(
```
Comment (mwydmuch) [inline pettingzoo_wrapper/observation_wrappers.py]: NIT: I think requiring cv2 just for resizing is a bit heavy. Maybe it should be replaced with Pillow, or just use pygame, which is already a dependency?

File: pettingzoo_wrapper/reward_wrappers.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+
+def _reset_info(info):
+    reset_info = info.get("reset_info")
+    if isinstance(reset_info, dict):
+        return reset_info
+    return None
+
+
+class HealthGatheringRewardWrapper(ParallelEnv):
```
Comment (mwydmuch) [inline pettingzoo_wrapper/reward_wrappers.py:14]: I think these should go to examples later.

File: pettingzoo_wrapper/base_pettingzoo_env_pipes.py [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+    pipe_end.send({"status": "initializing", "agent": agent})
+
+    try:
+        if not is_host:
+            import random
+
+            time.sleep(0.5 + random.uniform(0.5, 1.0))
+            game.add_game_args("+viz_connect_timeout 45")
+        game.init()
+        game.send_game_command("viz_respawn_delay 0")
```
Comment (mwydmuch) [inline pettingzoo_wrapper/base_pettingzoo_env_pipes.py:77]: This should ideally be part of an env config

File: pettingzoo_wrapper/__init__.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+# scenario-specific wrappers
+_WRAPPERS = {
+    "pitfall_multi_agent": PitfallRewardWrapper,
+    "remedy_rush_multi_agent": RemedyRushRewardWrapper,
+    "health_gathering_multi_agent": HealthGatheringRewardWrapper,
+}
+
+
+def make(
```
Comment (mwydmuch) [inline pettingzoo_wrapper/__init__.py:26]: In the feature we should have specific classes for each type of env

File: pettingzoo_wrapper/base_env_common.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    game.set_ticrate(ticrate)
+    game.set_mode(Mode.ASYNC_PLAYER if async_mode else Mode.PLAYER)
+    if timeout is not None:
+        game.set_episode_timeout(timeout)
+    if seed is not None:
+        game.set_seed(int(seed))
+    if is_host:
+        game.add_game_args(
+            f"-host {num_agents} -port {port} -netmode {netmode} "
+            "+timelimit 0 +sv_noautoaim 1 +sv_nocrouch 1 +sv_nofreelook 1 "
```
Comment (mwydmuch) [inline pettingzoo_wrapper/base_env_common.py:52]: `+timelimit 0 +sv_noautoaim 1 +sv_nocrouch 1 +sv_nofreelook 1 +sv_spawnfarthest 1 +sv_forcerespawn 1 +viz_respawn_delay 0` ideally should be part of config

File: pettingzoo_wrapper/observation_wrappers.py (hunk 3 above)
Comment (lmBored) [reply pettingzoo_wrapper/observation_wrappers.py]: Initially this was implemented as a manual numpy interpolation, but the performance doesn't change compared to using cv2, that's why I used cv2 here. Do you mean "heavy" as in disk space/more dependency? If so I can switch to `pygame`

**Review by mwydmuch — COMMENTED** [review]
Thank you @lmBored for improvments, this looks good. I made some initial tests and seems to run well. My proposal is to merge these and continue working on the wrapper to shape it more into pettingzoo conventions + add more tests and examples.

File: pettingzoo_wrapper/observation_wrappers.py (hunk 3 above)
Comment (mwydmuch) [reply pettingzoo_wrapper/observation_wrappers.py]: Yes, OpenCV is like a 70+ mb package, while Pillow or PyGame are just a few mb dependencies, and I'm pretty sure they resize in almost the same way.

**Comment by lmBored** [comment]
image

image

@mwydmuch I tried replacing opencv with raw numpy, Pillow, and Pygame, and throughput drops quite substantially depending on the tool. The green line is raw numpy resizing, gray line is OpenCV, brown is Pygame, purple is Pillow with `LANCZOS`, and the line with beach sand color is Pillow with `BILINEAR`. For context, I used `INTER_AREA` for opencv and `smoothscale` in Pygame.

Figure 2 is the learnability, which in my opinion is also important to take into account.

OpenCV is optimized for image resizing, etc. so this makes sense, but still a trade-off. Does this seem reasonable to you, to still switch to PyGame/pillow?

**Comment by mwydmuch** [comment]
Sorry for the long response time. I wonder if we actually need resizing, as ViZDoom allows us to control resolution, maybe additional resizing should be left to the external wrapper

**Comment by lmBored** [comment]
No worries. Resizing is a practice to increase performance, but I agree that we can leave it out. FYI this is performance of not scaling (the pink line at is around 1750 FPS); and white line is opencv resizing

Edit: And learnability improves, obviously

image
